### PR TITLE
[SPARK-31450][SQL] Make ExpressionEncoder thread-safe

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR moves the `ExpressionEncoder.toRow` and `ExpressionEncoder.fromRow` functions into their own function objects(`ExpressionEncoder.Serializer` & `ExpressionEncoder.Deserializer`). This effectively makes the `ExpressionEncoder` stateless, thread-safe and (more) reusable. The function objects are not thread safe, however they are documented as such and should be used in a more limited scope (making it easier to reason about thread safety).

### Why are the changes needed?
ExpressionEncoders are not thread-safe. We had various (nasty) bugs because of this.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Existing tests.

### CODE REVIEW

**Comment by hvanhovell** [comment]
Reviewers please first take a look at the `ExpressionEncoder` class and specifically the naming of the function objects.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 1]
```diff
@@ -384,7 +407,7 @@ case class ExpressionEncoder[T](
 
   override def toString: String = s"class[$schemaString]"
 
-  override def makeCopy: ExpressionEncoder[T] = copy()
+  override def makeCopy: ExpressionEncoder[T] = this
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:410]: do we still need this API?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:410]: It was added by https://issues.apache.org/jira/browse/SPARK-28456, in 3.0

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:410]: No. I can try to remove it, but we might break a 'public' API.

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:410]: Oh ok, if we port to 3.0 then it is safe to remove.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 2]
```diff
@@ -110,8 +110,8 @@ case class ScalaUDF(
     } else {
       val encoder = inputEncoders(i)
       if (encoder.isDefined && encoder.get.isSerializedAsStructForTopLevel) {
-        val enc = encoder.get.resolveAndBind()
-        row: Any => enc.fromRow(row.asInstanceOf[InternalRow])
+        val deserializer = encoder.get.resolveAndBind().createDeserializer()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:113]: shall we name it `fromRow` to be consistent with other places?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala [hunk 3]
```diff
@@ -445,8 +445,8 @@ class ObjectExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     testTypes.foreach { dt =>
       genSchema(dt).map { schema =>
         val row = RandomDataGenerator.randomRow(random, schema)
-        val rowConverter = RowEncoder(schema)
-        val internalRow = rowConverter.toRow(row)
+        val rowConverter = RowEncoder(schema).createSerializer()
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala:448]: `toRow`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/ArrayDataIndexedSeqSuite.scala [hunk 4]
```diff
@@ -73,8 +73,8 @@ class ArrayDataIndexedSeqSuite extends SparkFunSuite {
     arrayTypes.foreach { dt =>
       val schema = StructType(StructField("col_1", dt, nullable = false) :: Nil)
       val row = RandomDataGenerator.randomRow(random, schema)
-      val rowConverter = RowEncoder(schema)
-      val internalRow = rowConverter.toRow(row)
+      val rowConverter = RowEncoder(schema).createSerializer()
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/ArrayDataIndexedSeqSuite.scala:76]: `toRow`

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 5]
```diff
@@ -2965,9 +2965,8 @@ class Dataset[T] private[sql](
    */
   def toLocalIterator(): java.util.Iterator[T] = {
     withAction("toLocalIterator", queryExecution) { plan =>
-      // `ExpressionEncoder` is not thread-safe, here we create a new encoder.
-      val enc = resolvedEnc.copy()
-      plan.executeToIterator().map(enc.fromRow).asJava
+      val deserializer = resolvedEnc.createDeserializer()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2968]: `fromRow`

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 6]
```diff
@@ -3387,9 +3386,10 @@ class Dataset[T] private[sql](
         new JSONOptions(Map.empty[String, String], sessionLocalTimeZone))
 
       new Iterator[String] {
+        private val serializer = exprEnc.createSerializer()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:3389]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 7]
```diff
@@ -3649,9 +3649,8 @@ class Dataset[T] private[sql](
    * Collect all elements from a spark plan.
    */
   private def collectFromPlan(plan: SparkPlan): Array[T] = {
-    // `ExpressionEncoder` is not thread-safe, here we create a new encoder.
-    val enc = resolvedEnc.copy()
-    plan.executeCollect().map(enc.fromRow)
+    val deserializer = resolvedEnc.createDeserializer()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:3652]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 8]
```diff
@@ -162,6 +163,18 @@ object ExpressionEncoder {
       e4: ExpressionEncoder[T4],
       e5: ExpressionEncoder[T5]): ExpressionEncoder[(T1, T2, T3, T4, T5)] =
     tuple(Seq(e1, e2, e3, e4, e5)).asInstanceOf[ExpressionEncoder[(T1, T2, T3, T4, T5)]]
+
+  /**
+   * Function that deserializes an [[InternalRow]] into an object of type `T`. Instances of this
+   * class are not meant to be thread-safe.
+   */
+  abstract class Deserializer[T] extends (InternalRow => T)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:171]: Implementation note. I opted to go with abstract classes so we can get monomorphic call sites in many cases.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 9]
```diff
@@ -332,30 +344,67 @@ case class ExpressionEncoder[T](
...
+   * `serializer.apply(..)` are allowed to return the same actual [[InternalRow]] object.  Thus,
+   *  the caller should copy the result before making another call if required.
    */
-  def toRow(t: T): InternalRow = try {
-    inputRow(0) = t
-    extractProjection(inputRow)
-  } catch {
-    case e: Exception =>
-      throw new RuntimeException(s"Error while encoding: $e\n" +
+  def createSerializer(): Serializer[T] = new Serializer[T] {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:353]: This currently relies on use serializing the enclosing encoder as well. We technically don't need the entire encoder but only a couple fields. I could move this class into the companion object and just use the fields I need.

**Comment by dongjoon-hyun** [comment]
Thank you, @hvanhovell . Do we need a benchmark about this change?

**Comment by hvanhovell** [comment]
@dongjoon-hyun it should be a bit of a lateral move performance wise. The expensive bit is generating the code and compiling it, and we are definitely not avoiding that. It might be a bit quicker because it does not excessively copy the expression encoder (which is not for free because of the work which is done in the constructor). What would you like to see benchmarked?

**Comment by hvanhovell** [comment]
@dongjoon-hyun it should be a bit of a lateral move performance wise. The expensive bit is generating the code and compiling it, and we are definitely not avoiding that. It might be a bit quicker because it does not excessively copy the expression encoder (which is not for free because of the work which is done in the constructor). What would you like to see benchmarked?

**Comment by dongjoon-hyun** [comment]
Got it. Thank you for confirmation, @hvanhovell .

**Review by viirya — APPROVED** [review]
Looks like in a good shape already.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 10]
```diff
@@ -332,30 +344,67 @@ case class ExpressionEncoder[T](
...
+      constructProjection = SafeProjection.create(optimizedDeserializer)
+    }
+    initialize()
+
+    override def apply(row: InternalRow): T = try {
+      constructProjection(row).get(0, ObjectType(clsTag.runtimeClass)).asInstanceOf[T]
+    } catch {
+      case e: Exception =>
+        throw new RuntimeException(s"Error while decoding: $e\n" +
+          s"${deserializer.simpleString(SQLConf.get.maxToStringFields)}", e)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:401]: As pointed already, some fields like `deserializer` are in enclosing encoder, and so currently looks like we will serialize entire encoder? Actually we did serialize entire encoder currently but yea it is better we can get rid of unnecessary.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 11]
```diff
@@ -332,30 +344,67 @@ case class ExpressionEncoder[T](
...
+  def createSerializer(): Serializer[T] = new Serializer[T] {
+    @transient
+    private var inputRow: GenericInternalRow = _
+
+    @transient
+    private var extractProjection: UnsafeProjection = _
+
+    private def initialize(): Unit = {
+      inputRow = new GenericInternalRow(1)
+      extractProjection = GenerateUnsafeProjection.generate(optimizedSerializer)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:362]: why can't this two be `lazy val`? performance concerns?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 9 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:353]: yea that would be better

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:353]: I'm also thinking about if we just need to pass the original (de)serializer expressions and do the optimization inside `Serializer` and `Deserializer` lazily.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala [hunk 12]
```diff
@@ -437,7 +437,7 @@ class ExpressionEncoderSuite extends CodegenInterpretedPlanTest with AnalysisTes
   testOverflowingBigNumeric(new BigInteger("9" * 100), "java very big int")
 
   encodeDecodeTest("foo" -> 1L, "makeCopy") {
-    Encoders.product[(String, Long)].makeCopy.asInstanceOf[ExpressionEncoder[(String, Long)]]
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala:440]: we can remove this test as it was added to verify `makeCopy`

**Comment by HyukjinKwon** [comment]
cc also @JoshRosen

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala [hunk 13]
```diff
@@ -57,10 +57,6 @@ import org.apache.spark.sql.types._
  *   Encoders.bean(MyClass.class);
  * }}}
  *
- * == Implementation ==
- *  - Encoders are not required to be thread-safe and thus they do not need to use locks to guard
- *    against concurrent access if they reuse internal buffers to improve performance.
- *
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala:63]: Removing this means Encoders must be thread-safe? Do we need explicit comment for that?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 9 above)
Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:353]: There is some cost to the optimization. So I would like to do it only once.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 11 above)
Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:362]: Yeah, lazy vals are not free and the `extractProjection` is on the hot path.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala (hunk 13 above)
Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/Encoder.scala:63]: I think it should be thread-safe.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala (hunk 12 above)
Comment (hvanhovell) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala:440]: Done

**Review by cloud-fan — APPROVED** [review]

**Comment by srowen** [comment]
I feel like @JoshRosen looked at this too a while ago

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you, @hvanhovell and all.
Merged to master/3.0.

**Comment by HeartSaVioR** [comment]
Late LGTM.

**Pull request closed without merging**